app/controllers/api/functions.php

```diff
@@ -1290,3 +1296,234 @@
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'functions')
+    ->label('sdk.method', 'updateVariable')
+    ->label('sdk.description', '/docs/references/functions/update-variable.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_VARIABLE)
+    ->param('functionId', null, new UID(), 'Function unique ID.', false)
+    ->param('variableId', null, new UID(), 'Variable unique ID.', false)
+    ->param('key', null, new Text(255), 'Variable key. Max length: 255 chars.', true)
+    ->param('value', null, new Text(16384), 'Variable value. Max length: 16384 chars.', true)
+    ->inject('response')
+    ->inject('dbForProject')
+    ->action(function (string $functionId, string $variableId, ?string $key, ?string $value, Response $response, Database $dbForProject) {
+        if (empty($key) && empty($value)) {
+            throw new Exception(Exception::VARIABLE_MISSING_PAYLOAD, 'Missing key or value. Define at least one.');
+        }
+
+        $function = $dbForProject->getDocument('functions', $functionId);
+
+        if ($function->isEmpty()) {
+            throw new Exception(Exception::FUNCTION_NOT_FOUND, 'Function not found');
+        }
+
+        $variable = $dbForProject->findOne('variables', [
+            Query::equal('_uid', [$variableId]),
+            Query::equal('functionInternalId', [$function->getInternalId()]),
+        ]);
+
+        if ($variable === false || $variable->isEmpty()) {
+            throw new Exception(Exception::VARIABLE_NOT_FOUND, 'Variable not found');
+        }
+
+        $variable
+            ->setAttribute('key', $key ?? $variable->getAttribute('key'))
+            ->setAttribute('value', $value ?? $variable->getAttribute('value'))
+            ->setAttribute('search', implode(' ', [$variableId, $function->getId(), $key]))
+        ;
+
+        try {
+            $dbForProject->updateDocument('variables', $variable->getId(), $variable);
+        } catch (DuplicateException $th) {
+            throw new Exception(Exception::VARIABLE_ALREADY_EXISTS, 'Variable key already used.');
+        }
+
+        $dbForProject->deleteCachedDocument('functions', $function->getId());
+
+        $response->dynamic($variable, Response::MODEL_VARIABLE);
+    });
+
+App::delete('/v1/functions/:functionId/variables/:variableId')
+    ->desc('Delete Variable')
+    ->groups(['api', 'functions'])
+    ->label('scope', 'functions.write')
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'functions')
+    ->label('sdk.method', 'deleteVariable')
+    ->label('sdk.description', '/docs/references/functions/delete-variable.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_NOCONTENT)
+    ->label('sdk.response.model', Response::MODEL_NONE)
+    ->param('functionId', null, new UID(), 'Function unique ID.', false)
```

Comment:
All ID params should default to an empty string, otherwise when not passed, a type error will be thrown due to trying to call the action with `null` where a non-nullable `string` is expected. 

Looks like this might be a general issue not just in this controller

Reply:
General issue. Making myself note, will address it in a separate PR in future.